lib/puma/cluster/worker.rb

```diff
@@ -122,49 +132,111 @@ def run
 
-          stat_thread ||= Thread.new(@worker_write) do |io|
-            Puma.set_thread_name "stat pld"
-            base_payload = "#{PIPE_PING}#{Process.pid}"
-
-            while true
-              begin
-                b = server.backlog || 0
-                r = server.running || 0
-                t = server.pool_capacity || 0
-                m = server.max_threads || 0
-                rc = server.requests_count || 0
-                bt = server.busy_threads || 0
-                payload = %Q!#{base_payload}{ "backlog":#{b}, "running":#{r}, "pool_capacity":#{t}, "max_threads":#{m}, "requests_count":#{rc}, "busy_threads":#{bt} }\n!
-                io << payload
-              rescue IOError
-                Puma::Util.purge_interrupt_queue
-                break
-              end
-              sleep @options[:worker_check_interval]
-            end
-          end
+          make_sure_pinging(server)
+
           server_thread.join
         end
 
+        if @mold
+          set_proc_title(role: "mold")
+
+          Signal.trap("SIGTERM") do
+            @worker_write << "#{PIPE_EXTERNAL_TERM}#{Process.pid}\n" rescue nil
+            @fork_pipe.close
+          end
+
+          worker_pids = []
+          Signal.trap "SIGCHLD" do
+            wakeup! if worker_pids.reject! do |p|
+              Process.wait(p, Process::WNOHANG) rescue true
+            end
+          end
+
+          @config.run_hooks(:on_mold_promotion, index, @log_writer, @hook_data)
+
+          make_sure_pinging(server)
+          wakeup!
+
+          begin
+            while (idx = PipeProtocols::Fork.read_from(@fork_pipe))
+              worker_pids << pid = spawn_worker(idx)
+              @worker_write << "#{PIPE_FORK}#{pid}:#{idx}\n"
+              log "Forked worker #{idx} with pid #{pid}"
+            end
+          rescue StandardError => e
+            log "Fork pipe terminated with exception: #{e.inspect}"
+          end
+
+          @config.run_hooks(:on_mold_shutdown, index, @log_writer, @hook_data)
+        end
         # Invoke any worker shutdown hooks so they can prevent the worker
         # exiting until any background operations are completed
-        @config.run_hooks(:before_worker_shutdown, index, @log_writer, @hook_data)
+        @config.run_hooks(:before_worker_shutdown, index, @log_writer, @hook_data) unless @mold
       ensure
-        @worker_write << "#{PIPE_TERM}#{Process.pid}\n" rescue nil
-        @worker_write.close
+        @worker_write << "#{PIPE_TERM}#{Process.pid}\n"
       end
 
       private
 
+      def make_sure_pinging(server)
+        # if the stat thread died, join and replace it
+        if @thread && !@thread.alive?
+          begin
+            @thread.join rescue nil # just ignore exceptions here
+            @thread = nil
+          rescue => e
+            log "While joining stat thread rescued #{e.class}: #{e.message}"
+          end
+        end
+
+        @thread ||= Thread.new(@worker_write) do |io|
```

Comment:
Considering we read this ivar at the very start of the method, I think a more descriptive name would be useful:

```suggestion
        @stat_thread ||= Thread.new(@worker_write) do |io|
```